tests/mass_enroll_errorstates_test.php

```diff
@@ -0,0 +1,117 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Upgrade code for local_mass_enroll
+ *
+ * @package     local_mass_enroll
+ * @author Andrew Hancox <andrewdchancox@googlemail.com>
+ * @author Open Source Learning <enquiries@opensourcelearning.co.uk>
+ * @link https://opensourcelearning.co.uk
+ * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @copyright 2023, Andrew Hancox
+ */
+
+namespace local_mass_enroll;
+
+use advanced_testcase;
+use context_course;
+use csv_import_reader;
+
+/**
+ * Mass enrol unit tests
+ *
+ * @package   local_mass_enroll
+ * @copyright 2023, Andrew Hancox
+ * @author Andrew Hancox <andrewdchancox@googlemail.com>
+ * @author Open Source Learning <enquiries@opensourcelearning.co.uk>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class mass_enroll_errorstates_test extends advanced_testcase {
+
+    public function errorstates_provider(): array {
+        return [
+            'withblank' => [
+                'file' => 'errorstate_users_by_idnumber_withblank.csv',
+                'messagecontains' => 'First column cannot be blank'
+            ],
+            'withguest' => [
+                'file' => 'errorstate_users_by_idnumber_withguest.csv',
+                'messagecontains' => 'Cannot enrol the guest user'
+            ],
+            'deleted' => [
+                'file' => 'errorstate_users_by_idnumber_deleted.csv',
+                'messagecontains' => 'Cannot enrol deleted user'
+            ]
+        ];
+    }
+
+    /**
+     * Tests mass_enroll function
+     * @param string $fixturefile Name of .csv in test fixtures directory to use
+     * @param string $messagecontains Warning message that should appear
+     * @dataProvider errorstates_provider
+     * @covers \local_mass_enroll\mass_enroll
+     */
+    public function test_mass_enroll_blankidnumber(string $fixturefile, string $messagecontains) {
+        global $DB, $CFG;
+        require_once(__DIR__.'/../lib.php');
+        require_once($CFG->libdir . '/csvlib.class.php');
+
+        $this->resetAfterTest();
+        $course = $this->getDataGenerator()->create_course();
+
+        $users = [];
+
+        for ($i = 1; $i < 5; $i++) {
+            $users[$i] = $this->getDataGenerator()->create_user(['idnumber' => $i]);
+        }
+        $this->getDataGenerator()->create_user(['idnumber' => 'deleted']);
+
+        $DB->execute("update {user} set idnumber = 'guest' where username = 'guest'");
```

Comment:
Any reason these cant be `$DB->set_field('user', 'idnumber', 'guest', ['username' => 'guest']);`